src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

```diff
@@ -282,7 +284,10 @@ private async Task ProcessSends()
                 var isCompleted = result.IsCompleted;
                 if (!buffer.IsEmpty)
                 {
-                    await _sender.SendAsync(buffer);
+                    _sender = _socketSenderPool.Rent();
```

Comment:
I'm a little confused why we need to rent and return for each send? Can't we rent once at the beginning of ProcessSends and return after (and calling Reset accordingly).

Reply:
The entire purpose of the change is to avoid that. Today we have one per connection, this reduces it from one per connection to one per send operation, hence why we go from 5000 in the test below to 8.

Reply:
I see. So this reduces allocations with a pool as well as reducing the amount of time an object is used for.

If we have a limit of 1024, couldn't that be worse though when you get a sudden increase in traffic and then reduces? Is that something to be concerned about here?

Reply:
Realistically for that to happen, you would need to cause the TCP buffers in the kernel to fill up per connection by sending massive payloads and having the client not read them fast enough. The only way the number of concurrent operations exceeds the number of executing sends is when the sends themselves go async and this is very rare. 

That said, the SocketSender is about 300 bytes a pop and we default to a maximum of 16 IOQueues, if by some glitch we end up with 1024 entries per IOQueue then we'd have 4MB of unfreeable memory.